### DISCUSSION TITLE
Wait until provider ready?

### DISCUSSION THREAD

**Discussion by tslocke** [adonisjs/core#3022] [Help] [answered]
I'm trying to get some variables automatically available in the repl, to access my db easily (arangodb collections).

It works great using `Repl.addMethod` in `start/repl.ts` to create a helper.

I was hoping I could go a step further and use `Repl.ready` and have those variables present automatically. It works with a simple test, e.g.

```
Repl.ready(repl => {
  repl.server.context['foo'] = 123
})
```

But... the provider for my db collections is not ready at that point in the load sequence, so I get an E_IOC_LOOKUP_FAILED error trying to import.

Is there any way to register a callback or wait on a promise until a specific provider is loaded?

Or perhaps a way to change the load order in `.adonisrc.json`?

Thanks

**Comment by thetutlage** [comment, marked as answer]
Yeah, the ready method is not asynchronous for reasons, as we want REPL to get interactive as soon as possible. I believe having a method that sets those values is the way to go here. 

```ts
import Repl from '@ioc:Adonis/Addons/Repl'

Repl.addMethod('loadSomething', async (repl) => {
  const res = await someMethod()
  repl.server.context[key] = res
})
```

**Reply by tslocke** [reply]
Yeah that's what I have right now.

`ready` is synchronous, but it can still call an async method and just not wait on the result. In this case I think it's fine - it's only a dev mode convenience, and by the time you type your first statement in the repl it will be ready.

All I need is some way to register a notification for when a particular IOC component is ready. I guess I could hack something in the provider itself, but I'd rather that code not get polluted.

**Reply by tslocke** [reply]
Doh! Just realised this is as simple as `await import('@ioc:MyApp/MyComponent')`

Got all my helper vars automatically in my repl now : )
